### ISSUE TITLE
Features.Set<IHttpResponseBodyFeature> no longer calling provided stream

### ISSUE THREAD

**Issue by RickStrahl** [dotnet/aspnetcore#37954]
I have some middleware that takes over the Response stream using `IHttpResponseBodyFeature` in order to take over Response processing and inject some HTML into the body content. This has been working fine in .NET 5.0, but in 6.0 it looks like the passed stream is no longer called by the Feature code.

The relevant code that creates the HTTP feature is this:

```cs
private async Task HandleHtmlInjection(HttpContext context)
{
    var path = context.Request.Path.Value;

    // Use a custom StreamWrapper to rewrite output on Write/WriteAsync
    using (var filteredResponse = new ResponseStreamWrapper(context.Response.Body, context))
    {
        // Use new IHttpResponseBodyFeature for abstractions of pilelines/streams etc.
        // For 3.x this works reliably while direct Response.Body was causing random HTTP failures
        context.Features.Set<IHttpResponseBodyFeature>(new StreamResponseBodyFeature(filteredResponse));

        await _next(context);
    }

}
```
[code here](https://github.com/RickStrahl/Westwind.AspnetCore.LiveReload/blob/master/Westwind.AspnetCore.LiveReload/LiveReloadMiddleware.cs#L56)

The [ResponseStreamWrapper](https://github.com/RickStrahl/Westwind.AspnetCore.LiveReload/blob/master/Westwind.AspnetCore.LiveReload/ResponseStreamWrapper.cs) inherits from Stream and implements various stream members, but none of those ever get called in 6.0 - they were in 5.0.

Can anybody point me at some documentation of what may have changed or what the suggested changed approach for replacing body content is now? 

### Further technical details
```
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-rc.2.21505.57\

Host (useful for support):
  Version: 6.0.0-rc.2.21480.5
  Commit:  6b11d64e7e

.NET SDKs installed:
  3.1.414 [C:\Program Files\dotnet\sdk]
  5.0.208 [C:\Program Files\dotnet\sdk]
  5.0.402 [C:\Program Files\dotnet\sdk]
  6.0.100-rc.2.21505.57 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-rc.2.21480.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.2.21480.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-rc.2.21501.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

</details>

**Comment by RickStrahl** [comment]
I was called away yesterday before I was able to update and push my repo to the server with the 6.0 updates. I've done that now and the latest update has the 6.0 targeting and 6.0 sample project here:

https://github.com/rickstrahl/westwind.aspnetcore.livereload

The project has runnable projects for .NET 5.0 and 6.0. 5.0 works, 6.0 does not - looks like my `ResponseStreamWrapper` is instantiated as a Response Feature but none of the Write methods are ever called (only FlushAsync it looks like).

This project is a relatively small middleware component that intercepts all HTML page requests and injects a small script into the page to force the page to LiveReload on changes. Behavior is irrelevant here, the key item is that it injects a small bit of JavaScript into every HTML page when enabled:

```html
<!-- West Wind Live Reload -->
<script src="/__livereloadscript"></script>
<!-- End Live Reload -->
```

As mentioned I see calls to `FlushAsync()` firing (which occurs before the first write) happen **but only for non HTML pages**. HTML pages (or generated cshtml or view content) never ends up firing into any of my methods. Non-HTML content is just passed through to the base stream, but HTML requests it looks I never get to hook as I was able to in 5.0. 

It could be a behavior change that results in not firing `FlushAsync()` first but I have no other hook to reset the Response ContentLength.

**Comment by davidfowl** [comment]
I assume you can reproduce this on the command line with dotnet run right?

**Comment by RickStrahl** [comment]
Yup :smile: - thought about that that some sort of hot reload functionality might be interfering but fails with plain `dotnet run` as well.

**Comment by davidfowl** [comment]
Took a quick look at it seems like [this](https://github.com/RickStrahl/Westwind.AspnetCore.LiveReload/blob/62584e03ad2571ebed2664eb7dd759339bb78fb9/Westwind.AspnetCore.LiveReload/ResponseStreamWrapper.cs#L164-L167) is the issue. You need to move the code into `WriteAsync(ReadOnlyMemory<byte> buffer)`.

**Comment by davidfowl** [comment]
I think this fixes it https://github.com/RickStrahl/Westwind.AspnetCore.LiveReload/pull/56. Let me know.

**Comment by RickStrahl** [comment]
David - thanks so much for taking a look and sure enough that solves the problem.

Just for clarification the issue I overlooked is that this `WriteAsync()` overload  is now called in v6: `public override  ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)`  where previously the `byte[]` version was called. I overlooked the implementation of this overload and didn't hook up the logic required to intercept requests as my old code had.
